File: dotnet/UrlShortener.Core/Database/PostgreSqlDatabase.cs
using Npgsql;
using UrlShortener.Core.Interfaces;
using UrlShortener.Core.Models;

namespace UrlShortener.Core.Database;

public class PostgreSqlDatabase : IUrlDatabase
{
    private readonly string _connectionString;

    public PostgreSqlDatabase(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task InitializeAsync()
    {
        using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var command = new NpgsqlCommand(@"
            CREATE TABLE IF NOT EXISTS urls (
                id SERIAL PRIMARY KEY,
                short_code VARCHAR(50) UNIQUE NOT NULL,
                original_url TEXT NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )", connection);

        await command.ExecuteNonQueryAsync();
    }

    public async Task<UrlMapping> CreateUrlAsync(string shortCode, string originalUrl)
    {
        using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var command = new NpgsqlCommand(@"
            INSERT INTO urls (short_code, original_url) 
            VALUES (@shortCode, @originalUrl) 
            RETURNING id", connection);
        
        command.Parameters.AddWithValue("shortCode", shortCode);
        command.Parameters.AddWithValue("originalUrl", originalUrl);

        var id = await command.ExecuteScalarAsync();

        return new UrlMapping
        {
            Id = Convert.ToInt32(id),
            ShortCode = shortCode,
            OriginalUrl = originalUrl,
            CreatedAt = DateTime.UtcNow
        };
    }

    public async Task<UrlMapping?> GetUrlAsync(string shortCode)
    {
        using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var command = new NpgsqlCommand("SELECT * FROM urls WHERE short_code = @shortCode", connection);
        command.Parameters.AddWithValue("shortCode", shortCode);

        using var reader = await command.ExecuteReaderAsync();
        if (await reader.ReadAsync())
        {
            return new UrlMapping
            {
                Id = reader.GetInt32(0), // id
                ShortCode = reader.GetString(1), // short_code
                OriginalUrl = reader.GetString(2), // original_url
                CreatedAt = reader.GetDateTime(3) // created_at
            };
        }

        return null;
    }

    public async Task<List<UrlMapping>> GetAllUrlsAsync()
    {
        using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var urls = new List<UrlMapping>();
        var command = new NpgsqlCommand("SELECT * FROM urls ORDER BY created_at DESC", connection);

        using var reader = await command.ExecuteReaderAsync();
        while (await reader.ReadAsync())
        {
            urls.Add(new UrlMapping
            {
                Id = reader.GetInt32(0), // id
                ShortCode = reader.GetString(1), // short_code
                OriginalUrl = reader.GetString(2), // original_url
                CreatedAt = reader.GetDateTime(3) // created_at
            });
        }

        return urls;
    }

    public async Task<bool> UrlExistsAsync(string shortCode)
    {
        var url = await GetUrlAsync(shortCode);
        return url != null;
    }

    public Task CloseAsync()
    {
        // NpgsqlConnection implements IDisposable, connections are managed per request
        return Task.CompletedTask;
    }
}

File: node/api/src/db/PostgreSQLDatabase.js
const { Pool } = require('pg');
const DatabaseInterface = require('./DatabaseInterface');

class PostgreSQLDatabase extends DatabaseInterface {
  constructor(connectionString) {
    super();
    this.pool = new Pool({
      connectionString: connectionString,
      ssl: process.env.NODE_ENV === 'production' ? { rejectUnauthorized: false } : false
    });
  }

  async initialize() {
    try {
      // Create table if it doesn't exist
      await this.pool.query(`
        CREATE TABLE IF NOT EXISTS urls (
          id SERIAL PRIMARY KEY,
          short_code VARCHAR(50) UNIQUE NOT NULL,
          original_url TEXT NOT NULL,
          created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        )
      `);
    } catch (err) {
      throw new Error(`PostgreSQL initialization failed: ${err.message}`);
    }
  }

  async createUrl(shortCode, originalUrl) {
    try {
      const result = await this.pool.query(
        'INSERT INTO urls (short_code, original_url) VALUES ($1, $2) RETURNING id',
        [shortCode, originalUrl]
      );
      
      return { shortCode, originalUrl, id: result.rows[0].id };
    } catch (err) {
      throw new Error(`Failed to create URL: ${err.message}`);
    }
  }

  async getUrl(shortCode) {
    try {
      const result = await this.pool.query(
        'SELECT * FROM urls WHERE short_code = $1',
        [shortCode]
      );
      
      if (result.rows.length > 0) {
        const row = result.rows[0];
        return {
          shortCode: row.short_code,
          originalUrl: row.original_url,
          createdAt: row.created_at
        };
      }
      return null;
    } catch (err) {
      throw new Error(`Failed to get URL: ${err.message}`);
    }
  }

  async getAllUrls() {
    try {
      const result = await this.pool.query(
        'SELECT * FROM urls ORDER BY created_at DESC'
      );
      
      return result.rows.map(row => ({
        shortCode: row.short_code,
        originalUrl: row.original_url,
        createdAt: row.created_at
      }));
    } catch (err) {
      throw new Error(`Failed to get all URLs: ${err.message}`);
    }
  }

  async urlExists(shortCode) {
    const url = await this.getUrl(shortCode);
    return !!url;
  }

  async close() {
    await this.pool.end();
  }
}

module.exports = PostgreSQLDatabase;
